extensions/markdown-language-features/src/features/smartSelect.ts

```diff
@@ -0,0 +1,151 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { MarkdownEngine } from '../markdownEngine';
+import { TableOfContentsProvider } from '../tableOfContentsProvider';
+
+export default class MarkdownSmartSelect implements vscode.SelectionRangeProvider {
+
+	constructor(
+		private readonly engine: MarkdownEngine
+	) { }
+
+	public async provideSelectionRanges(document: vscode.TextDocument, positions: vscode.Position[], _token: vscode.CancellationToken): Promise<vscode.SelectionRange[]> {
+		return await Promise.all(positions.map(async (position) => {
+			return await this.provideSelectionRange(document, position, _token);
+		}));
+	}
+
+	private async provideSelectionRange(document: vscode.TextDocument, position: vscode.Position, _token: vscode.CancellationToken): Promise<vscode.SelectionRange> {
+		let ranges : (vscode.SelectionRange|undefined) [] = await Promise.all([
+			await this.getHeaderSelectionRanges(document, position),
+			await this.getBlockSelectionRanges(document, position)
+		]);
+		let result : (vscode.SelectionRange) [] = ranges.filter(range => range !== undefined) as vscode.SelectionRange[];
+		if (result.length === 2) {
+			let header = result[0];
+			let block = result[1];
+			let blockParent = block.parent;
+
+			if (blockParent) {
+				if (header.range.contains(blockParent.range)) {
+					let revisedParent = new vscode.SelectionRange(blockParent.range, header);
+					if (revisedParent.range.contains(block.range)) {
+						return new vscode.SelectionRange(block.range, revisedParent);
+					}
+				} else {
+					if (header.range.contains(block.range)) {
+						return new vscode.SelectionRange(block.range, header);
+					}
+				}
+			} else {
+				if (header.range.contains(block.range)) {
+					return new vscode.SelectionRange(block.range, header);
+				}
+			}
+			return block;
+		} else {
+			return result[0];
+		}
+	}
+
+	private async getBlockSelectionRanges(document: vscode.TextDocument, position: vscode.Position): Promise<vscode.SelectionRange| undefined> {
+
+		const tokens = await this.engine.parse(document);
+
+		let nearbyTokens = tokens.filter(token => token.map && (token.map[0] <= position.line && token.map[1] >= position.line));
+
+		// sort from smallest to largest line range
+		let sortedTokens = nearbyTokens.sort((tokenOne, tokenTwo) => (tokenTwo.map[1] - tokenTwo.map[0] - tokenOne.map[1] - tokenOne.map[0]));
+
+		if (sortedTokens.length === 0) {
+			return undefined;
+		}
+
+		let parentToken = sortedTokens.pop();
```

Comment:
I'm having trouble following this logic. I suggest that you try extracting some of the parts out . Once you start refactoring, you may find that the logic also become clearer and cleaner. Adding additional tests may also help clarify what the core logic should look like